server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -0,0 +1,130 @@
+import org.apache.james.core.MailAddress
+import org.apache.james.lifecycle.api.LifecycleUtil
+import org.apache.james.rate.limiter.api.{AcceptableRate, AllowedQuantity, RateExceeded, RateLimiter, RateLimiterFactory, RateLimiterFactoryProvider, RateLimitingKey, RateLimitingResult, Rule, Rules}
+import org.apache.james.util.DurationParser
+import org.apache.mailet.Mail
+import org.apache.mailet.base.GenericMailet
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import scala.jdk.CollectionConverters._
+import scala.util.Using;
+
+case class PerRecipientRateLimiter(rateLimiter: RateLimiter, keyPrefix: Option[KeyPrefix], entityType: EntityType) {
+  def rateLimit(recipient: MailAddress, mail: Mail): Publisher[RateLimitingResult] =
+    entityType.extractQuantity(mail)
+      .map(increment => rateLimiter.rateLimit(RecipientKey(keyPrefix, entityType, recipient), increment))
+      .getOrElse(SMono.just[RateLimitingResult](RateExceeded))
+}
+
+case class RecipientKey(keyPrefix: Option[KeyPrefix], entityType: EntityType, mailAddress: MailAddress) extends RateLimitingKey {
+  override def asString(): String = s"${
+    keyPrefix.map(prefix => prefix.value + "_")
+      .getOrElse("")
+  }${entityType.asString()}_${mailAddress.asString()}"
+}
+
+class PerRecipientRateLimitMailet @Inject()(rateLimiterFactoryProvider: RateLimiterFactoryProvider) extends GenericMailet {
+  private var exceededProcessor: String = _
+  private var rateLimiters: Seq[PerRecipientRateLimiter] = _
+
+  override def init(): Unit = {
+    val rateLimiterFactory: RateLimiterFactory = rateLimiterFactoryProvider.create(getMailetConfig)
+    val duration: Duration = parseDuration()
+    val keyPrefix: Option[KeyPrefix] = Option(getInitParameter("keyPrefix")).map(KeyPrefix)
+    exceededProcessor = getInitParameter("exceededProcessor", Mail.ERROR)
+
+    val countRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("count"))
+      .flatMap(parameter => extractCountToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Count)))
+
+    val sizeRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("size"))
+      .map {
+        case "" => throw new IllegalArgumentException("'size' field cannot be empty if specified")
+        case s => s
+      }
+      .flatMap(parameter => extractSizeToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Size)))
+
+    rateLimiters = Seq(countRateLimiter, sizeRateLimiter)
+      .filter(limiter => limiter.isDefined)
+      .map(limiter => limiter.get)
+  }
+
+  override def service(mail: Mail): Unit = {
+    if (!mail.getRecipients.isEmpty) {
+      val rateLimitResults: Seq[(MailAddress, RateLimitingResult)] = mail.getRecipients.asScala.toSeq
+        .map(recipient => (recipient, applyRateLimiter(mail, recipient)))
```

Comment:
Like this we are blocking for each recipients, that are processed sequentially.

Can we only block once, thus call the rate limitter for all recipients reactively and parralelize a bit this?